substrate/frame/staking/src/slashing.rs

```diff
@@ -321,23 +321,52 @@ fn kick_out_if_recent<T: Config>(params: SlashParams<T>) {
 }
 
 /// Inform the [`DisablingStrategy`] implementation about the new offender and disable the list of
-/// validators provided by [`make_disabling_decision`].
+/// validators provided by [`decision`].
 fn add_offending_validator<T: Config>(params: &SlashParams<T>) {
-	DisabledValidators::<T>::mutate(|disabled| {
-		if let Some(offender) =
-			T::DisablingStrategy::decision(params.stash, params.slash_era, &disabled)
-		{
-			// Add the validator to `DisabledValidators` and disable it. Do nothing if it is
-			// already disabled.
-			if let Err(index) = disabled.binary_search_by_key(&offender, |index| *index) {
-				disabled.insert(index, offender);
-				T::SessionInterface::disable_validator(offender);
-			}
-		}
-	});
-
-	// `DisabledValidators` should be kept sorted
-	debug_assert!(DisabledValidators::<T>::get().windows(2).all(|pair| pair[0] < pair[1]));
+    DisabledValidators::<T>::mutate(|disabled| {
+        let decision = T::DisablingStrategy::decision(params.stash, params.slash, params.slash_era, &disabled);
+
+        match (decision.disable, decision.reenable) {
```

Comment:
Probably I am missing something but won't this be simpler if the two values are handled separately?
I mean if `decision.disable` is `Some` - disable the guy. Then if `decision.reenable` is `Some` - reenable the guy?

That way you will avoid the code duplication.

Reply:
Initially opted for this approach as this was alliged with all the possible outputs of decision. Just for readability and explicit separate handling. Maybe it was an overkill so switched back to your suggestion.